### PR TITLE
feat: [IOCOM-689] Move Third Part Data retrieval before Message Details screen

### PR DESCRIPTION
**Pull request by Vangaorth**
## Short description
This PR moves the logic to retrieve a message in a dedicated saga, while showing the Message Router Screen

## List of changes proposed in this pull request
- added a new reducer to handle the status of message data loading/aggregation
- added a new saga that implements the old Message Router Screen data retrieval logic. It also takes care of downloading the third party data if needed (and it marks the message as read only if such third party data have been successfully retrieved - same logic for SEND messages will come later in another PR)
- Message Router Screen: refactored in order to be a functional component and to just render itself, taking care only of the navigation and of the initial action dispatch. It also shows an error if the saga fails
- Fixed a condition where the user could get to a message details from a push notification without having accepted the message's preconditions
- MessageDetailsScreen details component does not retrieve Third Party Data on its own anymore (since the saga takes care of that now)

## How to test
Using the io-dev-api-server, check that any kind of message can be properly retrieved (standard messages, messages with payments, third party data messages, EU Covid messages and SEND messages)

### CODE REVIEW

File: ts/features/messages/saga/messageData.ts [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+  if (!pot.isSome(initialMessagePot) || pot.isError(initialMessagePot)) {
+    yield* put(loadMessageById.request({ id: messageId }));
+
+    const outputAction = yield* take([
+      loadMessageById.success,
+      loadMessageById.failure
+    ]);
+    if (!isActionOf(loadMessageById.success, outputAction)) {
+      return undefined;
+    }
```
Comment (adelloste) [inline ts/features/messages/saga/messageData.ts:167]:
```suggestion
    if (isActionOf(loadMessageById.failure, outputAction)) {
      return undefined;
    }
```

File: ts/features/messages/saga/messageData.ts [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+      })
+    );
+
+    const outputAction = yield* take([
+      upsertMessageStatusAttributes.success,
+      upsertMessageStatusAttributes.failure
+    ]);
+    if (!isActionOf(upsertMessageStatusAttributes.success, outputAction)) {
+      return undefined;
+    }
```
Comment (adelloste) [inline ts/features/messages/saga/messageData.ts:251]: same of above
```suggestion
    if (isActionOf(upsertMessageStatusAttributes.failure, outputAction)) {
      return undefined;
    }
```

File: ts/screens/messages/MessageRouterScreen.tsx [hunk 3]
```diff
@@ -57,233 +33,118 @@ type NavigationProps = IOStackNavigationRouteProps<
...
+        navigation.navigate(ROUTES.MESSAGES_NAVIGATOR, {
+          screen: EUCOVIDCERT_ROUTES.MAIN,
+          params: {
+            screen: EUCOVIDCERT_ROUTES.CERTIFICATE,
+            params: {
+              authCode: data.euCovidCerficateAuthCode,
+              messageId: data.messageId
+            }
+          }
+        });
```
[the remark this answers is not in the document]
Comment (Vangaorth) [reply ts/screens/messages/MessageRouterScreen.tsx]: Applied

File: ts/features/messages/saga/messageData.ts
[the remark this answers is not in the document]
Comment (Vangaorth) [reply ts/features/messages/saga/messageData.ts:1]: Done it

File: ts/components/messages/MessageDetail/index.tsx [hunk 4]
```diff
@@ -1,4 +1,5 @@
-import React, { useCallback, useEffect, useState } from "react";
+import * as React from "react";
+import { useCallback, useState } from "react";
```
[the remark this answers is not in the document]
Comment (Vangaorth) [reply ts/components/messages/MessageDetail/index.tsx]: Done it

File: ts/features/messages/saga/messageData.ts [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+
+  yield* put(loadThirdPartyMessage.request(messageId));
+
+  const outputAction = yield* take([
+    loadThirdPartyMessage.success,
+    loadThirdPartyMessage.failure
+  ]);
+  if (!isActionOf(loadThirdPartyMessage.success, outputAction)) {
+    return undefined;
+  }
```
[the remark this answers is not in the document]
Comment (Vangaorth) [reply ts/features/messages/saga/messageData.ts]: Done it

File: ts/features/messages/saga/messageData.ts (hunk 2 above)
Comment (Vangaorth) [reply ts/features/messages/saga/messageData.ts]: Done it

File: ts/features/messages/saga/__test__/messageData.test.ts
Comment (adelloste) [inline ts/features/messages/saga/__test__/messageData.test.ts:1]: We should also rename this file to `handleLoadMessageData`

**Review by adelloste — APPROVED** [review]

**Pull request merged**